relay-general/src/user_agent.rs

```diff
@@ -26,27 +26,55 @@ static UA_PARSER: Lazy<UserAgentParser> = Lazy::new(|| {
     )
 });
 
-fn get_user_agent_from_headers(headers: &Headers) -> Option<&str> {
-    for item in headers.iter() {
-        if let Some((ref o_k, ref v)) = item.value() {
-            if let Some(k) = o_k.as_str() {
-                if k.to_lowercase() == "user-agent" {
-                    return v.as_str();
+/// This has both the user agent string and the client hints, useful for the scenarios where
+/// you will use either information from client hints if it exists and if not fall back to
+/// user agent string.
+#[derive(Default, Debug)]
+pub struct RawUserAgentInfo<'a> {
+    /// The "old style" of a single UA string.
+    pub user_agent: Option<&'a str>,
+    pub client_hints: ClientHints<'a>,
+}
+
+/// The client hint variable names mirror the name of the headers.
+/// '<https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers#user_agent_client_hints>'
+#[derive(Default, Debug)]
+pub struct ClientHints<'a> {
+    /// your OS, e.g. macos, android ..
+    pub sec_ch_ua_platform: Option<&'a str>,
+    /// the version number of your OS
+    pub sec_ch_ua_platform_version: Option<&'a str>,
+    /// name of web browser and its version
+    pub sec_ch_ua: Option<&'a str>,
+    /// device model, e.g. samsung galaxy 3
+    pub sec_ch_ua_model: Option<&'a str>,
```

Comment:
Now that these fields are in a struct separate from the user-agent header, we can drop the prefix:

```suggestion
    /// The client's OS, e.g. macos, android...
    pub platform: Option<&'a str>,
    /// The version number of the client's OS.
    pub platform_version: Option<&'a str>,
    /// Name of the client's web browser and its version.
    pub ua: Option<&'a str>,
    /// Device model, e.g. samsung galaxy 3.
    pub ua_model: Option<&'a str>,
```

Reply:
im assuming in the future when working on this we would go a lot back and forth between the specs and the code, wouldn't it be easier to maintain when the names are exactly mirrored, even if it's a bit verbose?